ooni/director.py

```diff
@@ -84,13 +91,69 @@ def __init__(self):
 
         self.failures = []
 
-        self.torControlProtocol = None
-
         # This deferred is fired once all the measurements and their reporting
         # tasks are completed.
         self.allTestsDone = defer.Deferred()
         self.sniffers = {}
 
+        self.input_store = input_store
+        self.deck_store = deck_store
+
+        self._reset_director_state()
+        self._reset_tor_state()
+
+        self._subscribers = []
+
+    def subscribe(self, handler):
+        self._subscribers.append(handler)
+
+    def unsubscribe(self, handler):
+        self._subscribers.remove(handler)
+
+    def notify(self, event):
+        for handler in self._subscribers:
+            handler(event)
```

Comment:
Does it make sense here to `try...catch` around calling `handler` such that all the subscribers are iterated even if some of them fail?